File: 12-ButtonLayout/ButtonLayout-OC/FZHButton.m
//
//  FZHButton.m
//  12-ButtonLayout
//
//  Created by 冯志浩 on 2017/6/20.
//  Copyright © 2017年 冯志浩. All rights reserved.
//

#import "FZHButton.h"

@implementation FZHButton

- (FZHButton *)initButtonWithButtonFrame:(CGRect)buttonFrame labelFrame:(CGRect)labelFrame imageViewFrame:(CGRect)imageViewFrame {
    if (self = [super initWithFrame:buttonFrame]) {
        //label
        _customLabel = [[UILabel alloc]initWithFrame:labelFrame];
        _customLabel.textAlignment = NSTextAlignmentCenter;
        
        //image
        _customImageView = [[UIImageView alloc]initWithFrame:imageViewFrame];
        
        [self addSubview:_customLabel];
        [self addSubview:_customImageView];
    }
    return self;
}

- (void)setTitle:(NSString *)title forState:(UIControlState)state {
    _customLabel.text = title;
}

- (void)setTitleColor:(UIColor *)color forState:(UIControlState)state {
    _customLabel.textColor = color;
}

- (void)setImage:(UIImage *)image forState:(UIControlState)state {
    _customImageView.image = image;
}
@end

File: 12-ButtonLayout/12-ButtonLayout/FZHButton.swift
//
//  FZHButton.swift
//  12-ButtonLayout
//
//  Created by 冯志浩 on 2017/6/20.
//  Copyright © 2017年 冯志浩. All rights reserved.
//

import UIKit

class FZHButton: UIButton {
    let customLabel = UILabel()
    let customImageView = UIImageView()
    
    init(buttonFrame: CGRect, labelFrame: CGRect, imageViewFrame: CGRect) {
        super.init(frame: buttonFrame)
        //Label
        customLabel.frame = labelFrame
        customLabel.textAlignment = .center
        
        //image
        customImageView.frame = imageViewFrame
        
        addSubview(customLabel)
        addSubview(customImageView)
    }
    
    override func setTitle(_ title: String?, for state: UIControlState) {
        customLabel.text = title
    }
    
    override func setTitleColor(_ color: UIColor?, for state: UIControlState) {
        customLabel.textColor = color
    }
    
    override func setImage(_ image: UIImage?, for state: UIControlState) {
        customImageView.image = image
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

}
